packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,123 @@
+
+export class AssociationManager {
+  public static associateProductWithPortfolio(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct) {
+    const associationKey = hashValues(portfolio.node.addr, product.node.addr, scope.stack.node.addr);
+    if (!this.associationMap.has(associationKey)) {
+      const association = new CfnPortfolioProductAssociation(scope, `PortfolioProductAssociation${associationKey}`, {
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+      });
+      this.associationMap.set(associationKey, new Map());
+      this.associationMap.get(associationKey).set(Constraints.ASSOCIATION, association);
+    }
+  }
+
+  public static addResourceUpdateConstraint(scope: cdk.Resource, product: IProduct, options: TagUpdatesOptions) {
+    const [pair, associationKey] = this.associationPrecheck(scope, product);
+    InputValidator.validateLength(this.generateAssocationString(scope, pair), 'description', 0, 2000, options.description);
+    if (!this.associationMap.get(associationKey).get(Constraints.RESOURCE_UPDATE)) {
+      const constraint = new CfnResourceUpdateConstraint(scope, `ResourceUpdateConstraint${associationKey}`, {
+        acceptLanguage: options.acceptLanguage,
+        description: options.description,
+        portfolioId: pair.portfolio.portfolioId,
+        productId: pair.product.productId,
+        tagUpdateOnProvisionedProduct: options.tagUpdateOnProvisionedProductAllowed || options.tagUpdateOnProvisionedProductAllowed === undefined
+          ? Allowed.ALLOWED : Allowed.NOT_ALLOWED,
+      });
+
+      constraint.addDependsOn(this.associationMap.get(associationKey).get(Constraints.ASSOCIATION));
+      this.associationMap.get(associationKey).set(Constraints.RESOURCE_UPDATE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple resource update constraints for association ${this.generateAssocationString(scope, pair)}`);
+    }
+  }
+
+  public static associateTagOption(scope: cdk.Resource, resourceId: string, tagOption: TagOption) {
+    Object.keys(tagOption).forEach(key => {
+      InputValidator.validateLength(resourceId, 'TagOption key', 1, 128, key);
+      tagOption[key].forEach(value => {
+        InputValidator.validateLength(resourceId, 'TagOption value', 1, 256, value.value);
+        const tagOptionKey = hashValues(key, value.value);
+        if (!this.tagOptionMap.has(tagOptionKey)) {
+          const tO = new CfnTagOption(scope, `TagOption${tagOptionKey}`, {
+            key: key,
+            value: value.value,
+            active: value.active ?? true,
+          });
+          this.tagOptionMap.set(tagOptionKey, tO);
+        }
+
+        new CfnTagOptionAssociation(scope, `TagOptionAssociation${hashValues(scope.node.addr, tagOptionKey)}`, {
+          resourceId: resourceId,
+          tagOptionId: this.tagOptionMap.get(tagOptionKey)!.ref,
+        });
+      });
+    });
+  }
+
+  private static associationMap = new Map<string, any>();
```

Comment:
1. Why the `any` here? Can't we narrow down the type here more?
2. Why `string`, not `Constraints`?

Reply:
I had hard time defining the type as a nested dict structure.  The `any` is another map with a `Constraint` enum as a key and a L1 Cfn as the value, that get's defined when an association is created.

Reply:
`Map<string, Map<Constrains, CfnResource>>`?

Reply:
The constraint is a key in the nested map.  The key at this level is the hash string.

Reply:
Map is simpler, but I suppose this could be defined as it's own class if necessary.

Reply:
> The constraint is a key in the nested map. The key at this level is the hash string.

So... `Map<string, Map<string, CfnClass1 | CfnClass2>>`?

Reply:
That works if you use a union type, will end up getting large when adding the other 4 in.

Reply:
How about adding in ```
type ConstraintResource = CfnPortfolioProductAssociation | CfnResourceUpdateConstraint;```  and then ```  private static associationMap = new Map<string, Map<Constraints, ConstraintResource>>();
```that way when its called here and in creating association there is less code

Reply:
Sure.

Reply:
added a type